lib/lambda_ethereum_consensus/state_transition/operations.ex

```diff
@@ -174,6 +174,74 @@ defmodule LambdaEthereumConsensus.StateTransition.Operations do
     Enum.reverse(withdrawals)
   end
 
+  @spec process_proposer_slashing(BeaconState.t(), SszTypes.ProposerSlashing.t()) ::
+          {:ok, BeaconState.t()} | {:error, String.t()}
+  def process_proposer_slashing(state, proposer_slashing) do
+    header_1 = proposer_slashing.signed_header_1.message
+    header_2 = proposer_slashing.signed_header_2.message
+    proposer = Enum.at(state.validators, header_1.proposer_index)
+
+    cond do
+      not (header_1.slot == header_2.slot) ->
+        {:error, "Slots doesn't match"}
+
+      not (header_1.proposer_index == header_2.proposer_index) ->
+        {:error, "Proposer indices doesn't match"}
+
+      not (header_1 != header_2) ->
+        {:error, "Headers are same"}
+
+      not Predicates.is_slashable_validator(proposer, Accessors.get_current_epoch(state)) ->
+        {:error, "Proposer is not slashable"}
+
+      true ->
+        is_verified =
+          [proposer_slashing.signed_header_1, proposer_slashing.signed_header_2]
+          |> Enum.reduce_while(false, fn signed_header, _acc ->
+            domain =
+              Accessors.get_domain(
+                state,
+                Constants.domain_beacon_proposer(),
+                Misc.compute_epoch_at_slot(signed_header.slot)
+              )
+
+            signing_root =
+              Misc.compute_signing_root(signed_header.message, domain)
+
+            bls_verify_proposer_slashing(
+              proposer.pubkey,
+              signing_root,
+              signed_header.signature
+            )
+          end)
+
+        if not is_verified do
+          {:error, "Signed header 1 or 2 signature is not verified"}
+        end
+
+        case Mutators.slash_validator(state, header_1.proposer_index) do
+          {:ok, state} -> {:ok, state}
+          {:error, msg} -> {:error, msg}
+        end
```

Comment:
The `case` isn't needed (since we just return the result in all cases), and it should go inside an `else` to make it work.

Reply:
I got the case part but was wondering in line 218 I am checking if the signature is verified and returning an error why doesn't it returns it as default and still executes 222

Reply:
Because [there's no "return" in Elixir](https://stackoverflow.com/questions/41520031/why-there-is-no-return-statement-in-elixir)

Reply:
yeap it doesn't but like lets say
```elixir
fn foo(x) do
  if x do
    true
  end
  false
end
```

returns false if x is true as well why this is the case

Reply:
Because, in Elixir, everything is an expression that evaluates to a value:
- `if` clauses evaluate to the last expression inside their `do` block's value if the condition is `true`, or the last expression inside their `else` block's otherwise (`nil` by default)
- function calls evaluate to the last expression in their `do` block's value
- literals evaluate to themselves

In this case, `foo(true)` evaluates to `false`, because

```elixir
if true do
  true
end
false
```
is equivalent to:
```elixir
true
false
```
Since the function returns only the last value, it returns `false`

Reply:
thanks ser!